services/dashboard/views/apps/app/appmaster.js

```diff
@@ -0,0 +1,120 @@
+/*
+ * Licensed under the Apache License, Version 2.0
+ * See accompanying LICENSE file.
+ */
+'use strict';
+angular.module('dashboard.apps.appmaster', ['directive.visgraph', 'dashboard.streamingdag'])
+
+  .config(['$routeProvider', function ($routeProvider) {
+    $routeProvider
+      .when('/apps/app/:id', {
+        label: 'Application ',
+        templateUrl: 'views/apps/app/appmaster.html',
+        controller: 'AppMasterCtrl'
+      });
+  }])
+
+  .controller('AppMasterCtrl',
+  ['$scope', '$routeParams', 'breadcrumbs', 'restapi', 'StreamingService', 'visdagUtil', 'StreamingDag',
+    function ($scope, $routeParams, breadcrumbs, restapi, StreamingService, visdagUtil, StreamingDag) {
+      $scope.app = {id: $routeParams.id};
+      breadcrumbs.options = {'Application ': 'Application ' + $scope.app.id};
+
+      $scope.streamingDag = null;
+      restapi.subscribe('/appmaster/' + $scope.app.id + '?detail=true', $scope,
+        function (data) {
+          // TODO: inactive application does not have this key
+          if (data.hasOwnProperty('appName')) {
+            $scope.app = {
+              actorPath: data.actorPath,
+              duration: data.clock,
+              executors: data.executors,
+              id: data.appId,
+              name: data.appName
+            };
+          }
+
+          // TODO: distributed shell does not have these two keys
+          if (data.hasOwnProperty('dag') && data.hasOwnProperty('processors')) {
+            if ($scope.streamingDag === null) {
+              $scope.streamingDag = new StreamingDag($scope.app.id, data.processors,
+                data.processorLevels, data.dag.edges);
+              $scope.visgraph = {
+                options: visdagUtil.newOptions(),
+                data: visdagUtil.newData()
+              };
+
+              // Usually metrics will be pushed by websocket. In worst case, metrics might be available
+              // in couple of seconds. This will cause some charts to be empty. For better user experience,
+              // we will manually fetch metrics via restapi at least once, before websocket is ready.
+              if ($scope.streamingDag.lastUpdated === null) {
+                restapi.repeatUntil('/metrics/app/' + $scope.app.id + '/app' + $scope.app.id, $scope,
```

Comment:
Add reference here to the data type HistoryMetrics. Something like:
//TODO datatype: HistoryMetrics. Try and do serde using scalajs See #458